pkg/bootstrap/instance_test.go

```diff
@@ -418,6 +419,80 @@ func checkOpencensusConfig(t *testing.T, got, want *bootstrap.Bootstrap) {
 	}
 }
 
+// Test that our golden files all have proper stat tags
+func checkStatsTags(t *testing.T, got *bootstrap.Bootstrap) {
+	for _, tag := range got.StatsConfig.StatsTags {
+		// TODO: Add checks for other tags
+		if tag.TagName == "cluster_name" {
+			checkClusterNameTag(t, tag.GetRegex())
+		}
+	}
+}
+
+// Envoy will remove the first capture group and set the tag to the second capture group.
+// We double check that the regex returns what we want here.
+func checkClusterNameTag(t *testing.T, regex string) {
+	if regex == "" {
+		t.Fatalf("cluster_name tag regex is empty")
+	}
+
+	compiledRegex, err := regexp.Compile(regex)
+	if err != nil {
+		t.Fatalf("invalid regex for cluster_name tag: %v", err)
+	}
+
+	tc := []struct {
+		name               string
+		clusterName        string
+		firstCaptureGroup  string
+		secondCaptureGroup string
+	}{
+		{
+			name:               "cluster_name stats tag - external service",
+			clusterName:        "cluster.outbound|443||api.facebook.com;.upstream_rq_retry",
+			firstCaptureGroup:  ".outbound|443||api.facebook.com;",
+			secondCaptureGroup: "outbound|443||api.facebook.com",
+		},
+		{
+			name:               "cluster_name stats tag - internal kubernetes service",
+			clusterName:        "cluster.outbound|443||kubernetes.default.svc.cluster.local;.upstream_rq_retry",
+			firstCaptureGroup:  ".outbound|443||kubernetes.default.svc.cluster.local;",
+			secondCaptureGroup: "outbound|443||kubernetes.default.svc.cluster.local",
+		},
+		{
+			name:               "cluster_name stats tag - no dots in cluster name",
+			clusterName:        "cluster.xds-grpc;.upstream_rq_retry",
+			firstCaptureGroup:  ".xds-grpc;",
+			secondCaptureGroup: "xds-grpc",
+		},
+	}
+
+	for _, tt := range tc {
+		t.Run(tt.name, func(t *testing.T) {
+			subMatches := compiledRegex.FindStringSubmatch(tt.clusterName)
+			if subMatches == nil {
+				t.Fatalf("cluster_name tag regex does not match cluster name %s", tt.clusterName)
+			}
+
+			// The first index is the whole match followed by N number of capture groups.
+			// There are 2 capture groups we expect in the regex, so we always check for 3
```

Comment:
Love seeing comments like this, actually explaining the magic number. Very helpful!